tests/src/test/java/alluxio/client/FreeAndDeleteIntegrationTest.java

```diff
@@ -0,0 +1,126 @@
+  public final void before() throws Exception {
+    mFileSystem = mLocalAlluxioClusterResource.get().getClient();
+    Configuration workerConfiguration = mLocalAlluxioClusterResource.get().getWorkerConf();
+    mWriteBoth = StreamOptionUtils.getCreateFileOptionsCacheThrough(workerConfiguration);
+  }
+
+  @BeforeClass
+  public static void beforeClass() {
+    HeartbeatContext.setTimerClass(HeartbeatContext.WORKER_BLOCK_SYNC,
+        HeartbeatContext.SCHEDULED_TIMER_CLASS);
+    HeartbeatContext.setTimerClass(HeartbeatContext.MASTER_LOST_FILES_DETECTION,
+        HeartbeatContext.SCHEDULED_TIMER_CLASS);
+  }
+
+  @AfterClass
+  public static void afterClass() {
+    HeartbeatContext.setTimerClass(HeartbeatContext.WORKER_BLOCK_SYNC,
+        HeartbeatContext.SLEEPING_TIMER_CLASS);
+    HeartbeatContext.setTimerClass(HeartbeatContext.MASTER_LOST_FILES_DETECTION,
+        HeartbeatContext.SLEEPING_TIMER_CLASS);
+  }
+
+  @Test
+  public void freeAndDeleteIntegrationTest() throws Exception {
+    AlluxioURI filePath = new AlluxioURI(PathUtils.uniqPath());
+    FileOutStream os = mFileSystem.createFile(filePath, mWriteBoth);
+    os.write((byte) 0);
+    os.write((byte) 1);
+    os.close();
+
+    URIStatus status = mFileSystem.getStatus(filePath);
+    Assert.assertEquals(PersistenceState.PERSISTED.toString(), status.getPersistenceState());
+
+    mFileSystem.free(filePath);
+    // Execute the blocks free, which needs two heartbeats
+    Assert.assertTrue(HeartbeatScheduler.await(HeartbeatContext.WORKER_BLOCK_SYNC, 5,
+        TimeUnit.SECONDS));
+    HeartbeatScheduler.schedule(HeartbeatContext.WORKER_BLOCK_SYNC);
+    Assert.assertTrue(HeartbeatScheduler.await(HeartbeatContext.WORKER_BLOCK_SYNC, 5,
+        TimeUnit.SECONDS));
+    HeartbeatScheduler.schedule(HeartbeatContext.WORKER_BLOCK_SYNC);
+    Assert.assertTrue(HeartbeatScheduler.await(HeartbeatContext.WORKER_BLOCK_SYNC, 5,
+        TimeUnit.SECONDS));
+
+    status = mFileSystem.getStatus(filePath);
+    Assert.assertEquals(PersistenceState.PERSISTED.toString(), status.getPersistenceState());
+    // Block metadata is still present after block freed.
+    Assert.assertEquals(1, status.getBlockIds().size());
+
+    mFileSystem.delete(filePath);
+    // File is immediately gone after delete.
+    mThrown.expect(FileDoesNotExistException.class);
+    mFileSystem.getStatus(filePath);
+
+    // Execute the lost files detection.
+    Assert.assertTrue(HeartbeatScheduler.await(HeartbeatContext.MASTER_LOST_FILES_DETECTION, 5,
+        TimeUnit.SECONDS));
+    HeartbeatScheduler.schedule(HeartbeatContext.MASTER_LOST_FILES_DETECTION);
+    Assert.assertTrue(HeartbeatScheduler.await(HeartbeatContext.MASTER_LOST_FILES_DETECTION, 5,
+        TimeUnit.SECONDS));
+
+    // Verify the blocks are not in mLostBlocks.
+    BlockMaster bm = PrivateAccess.getBlockMaster(mLocalAlluxioClusterResource.get().getMaster()
+        .getInternalMaster());
+    Assert.assertEquals(0, bm.getLostBlocks().size());
```

Comment:
Added, thanks for the suggestion! Actually identified a problem in setting up testing environment by adding this check.